presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1288,25 +1349,97 @@ public void close()
         {
             try {
-                log.debug("Starting upload for host: %s, key: %s, file: %s, size: %s", host, key, tempFile, tempFile.length());
+                if (!upload.isPresent()) {
+                    InitiateMultipartUploadRequest request = new InitiateMultipartUploadRequest(host, key)
+                            .withStorageClass(s3StorageClass)
+                            .withCannedACL(aclType);
+                    if (sseEnabled) {
+                        switch (sseType) {
+                            case KMS:
+                                if (sseKmsKeyId != null) {
+                                    request.withSSEAwsKeyManagementParams(new SSEAwsKeyManagementParams(sseKmsKeyId));
+                                }
+                                else {
+                                    request.withSSEAwsKeyManagementParams(new SSEAwsKeyManagementParams());
+                                }
+                                break;
+                            case S3:
+                                ObjectMetadata metadata = new ObjectMetadata();
+                                metadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+                                request.withObjectMetadata(metadata);
+                                break;
+                        }
+                    }
+                    upload = Optional.of(client.initiateMultipartUpload(request));
+                }
+                InitiateMultipartUploadResult currentUpload = upload.get();
+                out.close();
                 STATS.uploadStarted();
+                UploadPartRequest request = new UploadPartRequest()
+                        .withBucketName(host)
+                        .withKey(key)
+                        .withFile(partFile.toFile())
+                        .withPartNumber(currentPart)
+                        .withFileOffset(0)
+                        .withLastPart(last)
+                        .withPartSize(partFile.toFile().length())
+                        .withUploadId(currentUpload.getUploadId());
+                if (log.isDebugEnabled()) {
+                    request.withGeneralProgressListener(createProgressListener(partFile.toFile().length()));
+                }
 
-                PutObjectRequest request = new PutObjectRequest(host, key, tempFile);
+                partList.add(client.uploadPart(request));
+                STATS.uploadSuccessful();
+                Files.deleteIfExists(partFile);
+                if (!last) {
+                    newTempFileBuffer();
+                }
+            }
+            catch (Exception e) {
+                STATS.uploadFailed();
+                upload.ifPresent(this::cleanUpFailedMultipartUpload);
+                throw new RuntimeException(e);
+            }
+        }
+
+        private String uploadSingleFile()
+        {
+            TransferManager transfer = TransferManagerBuilder.standard()
```

Comment:
It could also be that the min part size was configured to a very large size. I didn't dig into what optimizations the transfer manager might make for larger-sized files. I just deferred to that implementation since it seems robust enough to handle large file uploads

Reply:
With the results [here](https://github.com/prestodb/presto/pull/22424#issuecomment-2052150813), I am of the opinion that we should let this PR merge since the performance is always better than master in the benchmark configurations. If we want to improve perf further, we can iterate in another PR on a proper streaming implementation. Trino also has a proper streaming implementation.

When implementing the pipelining I think there are more design decisions that we may wish to make (or further configuration) which is not really in scope for the goal of the PR, which was lifting the disk space limitation.